File: Maximum Number of Points with Cost.go
func maxPoints(points [][]int) int64 {
    n := len(points)
    m := len(points[0])
    for i := 0; i < n-1; i++ {
        maxx := 0
        dp := make([]int, m)
        for j := 0; j < m; j++ {
            maxx = max(maxx-1, points[i][j])
            dp[j] = maxx
        }
        maxx = 0
        for j := m-1; j > -1; j-- {
            maxx = max(maxx-1, points[i][j])
            points[i+1][j] += max(dp[j], maxx)
        }
    }
    var ans int64
    for _, val := range points[n-1] {
        ans = max(ans, int64(val))
    }
    return ans
}

File: Maximum Number of Points with Cost.py
class Solution:
    def maxPoints(self, points: List[List[int]]) -> int:
        n = len(points)
        m = len(points[0])
        for i in range(n-1):
            maxx = 0
            dp = [0] * m
            for j in range(m):
                maxx = max(maxx-1, points[i][j])
                dp[j] = max(dp[j], maxx)

            maxx = 0
            for j in range(m-1, -1, -1):
                maxx = max(maxx-1, points[i][j])
                dp[j] = max(dp[j], maxx)
                points[i+1][j] += dp[j]
        return max(points[-1])
